File: go/internal/domain/sku.go
package domain

import (
	"fmt"
	"strings"
)

// SkuMaxUTF8Bytes is the shared inclusive UTF-8 byte length limit for a SKU.
const SkuMaxUTF8Bytes = 64

// skuEdgeWhitespace is the only set of characters stripped from SKU ends:
// ASCII space, tab, CR, and LF. U+00A0 and other non-ASCII whitespace stay.
const skuEdgeWhitespace = " \t\r\n"

// Sku is a non-empty stock-keeping-unit code, normalized to its trimmed form.
type Sku string

// NewSku strips ASCII space/tab/CR/LF from both ends and rejects codes that
// end up empty or exceed the UTF-8 byte limit. Interior text and case are
// preserved; NBSP is not stripped.
func NewSku(code string) (Sku, error) {
	trimmed := strings.Trim(code, skuEdgeWhitespace)
	if trimmed == "" {
		return "", fmt.Errorf("new sku: %w", InvalidOrder{Reason: "sku code must be non-empty"})
	}
	if len(trimmed) > SkuMaxUTF8Bytes {
		return "", fmt.Errorf(
			"new sku: %w",
			InvalidOrder{Reason: fmt.Sprintf("sku code exceeds %d UTF-8 bytes", SkuMaxUTF8Bytes)},
		)
	}
	return Sku(trimmed), nil
}

// MustSku is NewSku for infallible literals; it panics on violation and is
// intended for tests and wiring code with compile-time-known inputs.
func MustSku(code string) Sku {
	sku, err := NewSku(code)
	if err != nil {
		panic(err)
	}
	return sku
}

File: swift/Sources/Warehouse/Domain/Sku.swift
/// Shared UTF-8 byte length limit for a normalized SKU code.
public let skuMaxUtf8Bytes = 64

/// A stock-keeping-unit code, normalized on creation.
public struct Sku: Equatable, Hashable, Sendable, CustomStringConvertible {
  public let code: String

  /// Strip only ASCII space, tab, CR, and LF from both ends, then validate.
  public init(_ code: String) throws {
    let trimmed = trimAsciiEdges(code)
    if trimmed.isEmpty {
      throw InvalidOrder("sku code must be non-empty")
    }
    if trimmed.utf8.count > skuMaxUtf8Bytes {
      throw InvalidOrder("sku code exceeds \(skuMaxUtf8Bytes) UTF-8 bytes")
    }
    self.code = trimmed
  }

  public var description: String { code }
}

func trimAsciiEdges(_ code: String) -> String {
  let edge: Set<Character> = [" ", "\t", "\r", "\n"]
  var start = code.startIndex
  var end = code.endIndex
  while start < end, edge.contains(code[start]) {
    start = code.index(after: start)
  }
  while end > start {
    let previous = code.index(before: end)
    if !edge.contains(code[previous]) {
      break
    }
    end = previous
  }
  return String(code[start..<end])
}
